File: android/app/src/main/java/com/nauhaan/skycast/ui/detail/DayDetailScreen.kt
package com.nauhaan.skycast.ui.detail

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.CloudOff
import androidx.compose.material.icons.filled.EventBusy
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.nauhaan.skycast.R
import com.nauhaan.skycast.core.designsystem.component.EmptyStateView
import com.nauhaan.skycast.core.designsystem.component.ErrorView
import com.nauhaan.skycast.core.designsystem.component.LoadingView
import com.nauhaan.skycast.core.designsystem.component.WeatherConditionBadge
import com.nauhaan.skycast.core.designsystem.theme.SkyCastTheme
import com.nauhaan.skycast.core.designsystem.theme.Spacing
import com.nauhaan.skycast.domain.model.ForecastDay
import com.nauhaan.skycast.domain.model.HourlyForecast
import com.nauhaan.skycast.domain.model.TemperatureUnit
import com.nauhaan.skycast.domain.model.WeatherCondition
import com.nauhaan.skycast.domain.model.WindSpeedUnit
import com.nauhaan.skycast.ui.common.toPresentation
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import kotlin.math.roundToInt

/** The 3-hourly breakdown for one forecast day, pushed from the Forecast tab. */
@Composable
fun DayDetailScreen(
    onNavigateBack: () -> Unit,
    modifier: Modifier = Modifier,
    viewModel: DayDetailViewModel = hiltViewModel(),
) {
    val uiState by viewModel.uiState.collectAsStateWithLifecycle()

    DayDetailContent(uiState = uiState, onNavigateBack = onNavigateBack, modifier = modifier)
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun DayDetailContent(uiState: DayDetailUiState, onNavigateBack: () -> Unit, modifier: Modifier = Modifier) {
    // Same format as the forecast row that pushed this screen. A different one there and here
    // reads as two different days.
    val title = uiState.date?.format(DAY_TITLE_FORMAT) ?: stringResource(R.string.detail_day_title)

    Scaffold(
        modifier = modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = { Text(title) },
                navigationIcon = {
                    IconButton(onClick = onNavigateBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = stringResource(R.string.action_back),
                        )
                    }
                },
            )
        },
    ) { innerPadding ->
        when {
            uiState.isMissing -> EmptyStateView(
                title = stringResource(R.string.day_detail_missing_title),
                message = stringResource(R.string.day_detail_missing_message),
                icon = Icons.Filled.EventBusy,
                modifier = Modifier.padding(innerPadding),
            )

            uiState.day == null && uiState.error != null -> {
                val presentation = uiState.error.toPresentation()
                ErrorView(
                    title = stringResource(presentation.titleRes),
                    message = stringResource(presentation.messageRes),
                    icon = Icons.Filled.CloudOff,
                    // No retry: the Forecast tab owns refreshing, and this screen has no
                    // location to refresh against until one resolves.
                    modifier = Modifier.padding(innerPadding),
                )
            }

            uiState.day == null -> LoadingView(modifier = Modifier.padding(innerPadding))

            else -> LazyColumn(modifier = Modifier.padding(innerPadding)) {
                item {
                    DaySummary(
                        locationName = uiState.locationName,
                        day = uiState.day,
                        unit = uiState.preferences.temperatureUnit,
                        modifier = Modifier.padding(Spacing.md),
                    )
                }
                item {
                    Text(
                        text = stringResource(R.string.day_detail_hourly_heading),
                        style = MaterialTheme.typography.titleMedium,
                        modifier = Modifier.padding(
                            start = Spacing.md,
                            end = Spacing.md,
                            bottom = Spacing.sm,
                        ),
                    )
                }
                items(uiState.day.hourly, key = { it.time.epochSecond }) { hour ->
                    HourlyRow(
                        hour = hour,
                        zoneOffset = uiState.zoneOffset,
                        temperatureUnit = uiState.preferences.temperatureUnit,
                        windUnit = uiState.preferences.windSpeedUnit,
                    )
                    HorizontalDivider()
                }
            }
        }
    }
}

/** The day's headline: condition, description and high/low. */
@Composable
private fun DaySummary(locationName: String, day: ForecastDay, unit: TemperatureUnit, modifier: Modifier = Modifier) {
    val high = unit.convertFromCelsius(day.maxTemperatureCelsius).roundToInt()
    val low = unit.convertFromCelsius(day.minTemperatureCelsius).roundToInt()
    val highLow = stringResource(R.string.day_detail_high_low, high, low, unit.symbol)

    Column(
        modifier = modifier
            .fillMaxWidth()
            .clearAndSetSemantics {
                contentDescription = "$locationName. ${day.description}. $highLow"
            },
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.xs),
    ) {
        Text(
            text = locationName,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        WeatherConditionBadge(
            condition = day.condition,
            // A day summary covers a whole day, so daytime artwork is the honest choice.
            isDaytime = true,
            modifier = Modifier.padding(vertical = Spacing.sm),
        )
        Text(
            text = day.description,
            // Emphasized type, on one element only: the screen's anchor.
            style = MaterialTheme.typography.titleLargeEmphasized,
            textAlign = TextAlign.Center,
        )
        Text(text = highLow, style = MaterialTheme.typography.bodyLarge)
    }
}

/** One 3-hourly reading. */
@Composable
private fun HourlyRow(
    hour: HourlyForecast,
    zoneOffset: ZoneOffset,
    temperatureUnit: TemperatureUnit,
    windUnit: WindSpeedUnit,
    modifier: Modifier = Modifier,
) {
    // The forecast location's clock, not the device's, otherwise a Maldivian forecast read from
    // London lists its afternoon readings as morning ones.
    val localTime = hour.time.atOffset(zoneOffset)
    val time = TIME_FORMAT.format(localTime)
    val temperature = temperatureUnit.convertFromCelsius(hour.temperatureCelsius).roundToInt()
    val wind = windUnit.convertFromMetresPerSecond(hour.windSpeedMetresPerSecond)
    val windText = stringResource(
        R.string.day_detail_wind,
        (wind * ONE_DECIMAL_SCALE).roundToInt() / ONE_DECIMAL_SCALE,
        windUnit.symbol,
    )
    val rain = (hour.precipitationProbability * PERCENT).roundToInt()

    // The forecast carries no sunrise/sunset, so daylight is approximated by clock hour. Getting
    // this wrong shows a sun at 3 am, the exact parity bug WeatherConditionIconTest guards
    // against, so it is worth being explicit rather than passing isDaytime = true.
    val isDaytime = localTime.hour in DAWN_HOUR until DUSK_HOUR

    val announcement = buildString {
        append(stringResource(R.string.day_detail_hour_accessibility, time, temperature, temperatureUnit.symbol))
        append(", ")
        append(windText)
        if (rain > 0) {
            append(", ")
            append(stringResource(R.string.day_detail_rain_accessibility, rain))
        }
    }

    Row(
        modifier = modifier
            .fillMaxWidth()
            .clearAndSetSemantics { contentDescription = announcement }
            .padding(horizontal = Spacing.md, vertical = Spacing.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            text = time,
            style = MaterialTheme.typography.bodyMedium,
            // A fixed width keeps the temperature column aligned down the list.
            modifier = Modifier.width(TimeColumnWidth),
        )
        WeatherConditionBadge(condition = hour.condition, isDaytime = isDaytime, size = HourBadgeSize)
        Text(
            text = "$temperature${temperatureUnit.symbol}",
            style = MaterialTheme.typography.titleMedium,
            modifier = Modifier.padding(start = Spacing.md),
        )
        Column(
            modifier = Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.End,
        ) {
            Text(
                text = windText,
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
            if (rain > 0) {
                Text(
                    text = "$rain%",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.primary,
                )
            }
        }
    }
}

private val TIME_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm")
private val DAY_TITLE_FORMAT: DateTimeFormatter = DateTimeFormatter.ofPattern("EEEE d MMM")
private const val PERCENT = 100
private const val ONE_DECIMAL_SCALE = 10.0
private const val DAWN_HOUR = 6
private const val DUSK_HOUR = 20
private val TimeColumnWidth = 56.dp
private val HourBadgeSize = 36.dp

/** Fixed date for the previews below: the longest day, so the hourly list is at its fullest. */
private val PreviewDate: LocalDate = LocalDate.parse("2026-06-21")

@Preview(name = "Day detail", showBackground = true)
@Composable
private fun DayDetailPreview() {
    SkyCastTheme {
        DayDetailContent(
            uiState = DayDetailUiState(
                locationName = "London",
                date = PreviewDate,
                day = previewDay(),
                isLoading = false,
            ),
            onNavigateBack = {},
        )
    }
}

private fun previewDay(): ForecastDay = ForecastDay(
    date = PreviewDate,
    condition = WeatherCondition.CLOUDS,
    description = "Scattered clouds",
    iconCode = "03d",
    minTemperatureCelsius = 14.0,
    maxTemperatureCelsius = 23.0,
    precipitationProbability = 0.2,
    hourly = listOf(
        HourlyForecast(
            time = Instant.parse("2026-06-21T03:00:00Z"),
            condition = WeatherCondition.CLEAR,
            iconCode = "01n",
            temperatureCelsius = 14.5,
            precipitationProbability = 0.0,
            windSpeedMetresPerSecond = 2.1,
        ),
        HourlyForecast(
            time = Instant.parse("2026-06-21T12:00:00Z"),
            condition = WeatherCondition.CLOUDS,
            iconCode = "03d",
            temperatureCelsius = 22.8,
            precipitationProbability = 0.2,
            windSpeedMetresPerSecond = 4.6,
        ),
    ),
)

File: ios/SkyCast/Features/Detail/DayDetailScreen.swift
import Foundation
import Observation
import SwiftUI

/// State for the pushed day-detail screen.
struct DayDetailUiState: Equatable {
    var locationName = ""
    var day: ForecastDay?
    /// The forecast location's zone, so the hourly rows read as that place's clock. UTC only for
    /// the initial, dataless state.
    var timeZone: TimeZone = .gmt
    var preferences: UserPreferences = .init()
    var isLoading = true
    var error: AppError?
    /// The requested day is not in the cached forecast, it has rolled out of the 5-day window.
    var isMissing = false
}

/// The 3-hourly breakdown for one forecast day.
///
/// Resolves the day from the cached forecast rather than taking a `ForecastDay` through the
/// navigation path, for the same reason ``LocationDetailViewModel`` takes an id: a value embedded
/// in the path is a snapshot that silently goes stale.
@MainActor
@Observable
final class DayDetailViewModel {
    private(set) var state = DayDetailUiState()

    private let locationID: Int64
    private let date: Date
    private let weatherRepository: any WeatherRepository
    private let locationRepository: any LocationRepository
    private let settingsStore: SettingsStore

    private var observationTask: Task<Void, Never>?

    init(
        locationID: Int64,
        date: Date,
        weatherRepository: any WeatherRepository,
        locationRepository: any LocationRepository,
        settingsStore: SettingsStore
    ) {
        self.locationID = locationID
        self.date = date
        self.weatherRepository = weatherRepository
        self.locationRepository = locationRepository
        self.settingsStore = settingsStore
    }

    func start() {
        guard observationTask == nil else { return }
        observationTask = Task { await observe() }
    }

    func stop() {
        observationTask?.cancel()
        observationTask = nil
    }

    private func observe() async {
        state.preferences = settingsStore.preferences
        do {
            guard let location = try await locationRepository.location(id: locationID) else {
                state.isLoading = false
                state.isMissing = true
                return
            }
            state.locationName = location.name

            for await dataState in weatherRepository.forecast(for: location) {
                if Task.isCancelled {
                    return
                }
                if let forecast = dataState.data {
                    state.locationName = forecast.locationName
                    state.timeZone = forecast.timeZone
                    // Compare by calendar day, not by instant: the date arrived from a row whose
                    // timestamp is the start of that day in the *device's* calendar, and an
                    // equality check on `Date` would fail on any difference at all.
                    // Compared in the forecast location's calendar, matching the zone the
                    // mapper grouped the days in. `Calendar.current` here would disagree with the
                    // grouping whenever the device and the place are in different zones.
                    var calendar = Calendar(identifier: .gregorian)
                    calendar.timeZone = forecast.timeZone
                    let match = forecast.days.first {
                        calendar.isDate($0.date, inSameDayAs: date)
                    }
                    state.day = match ?? state.day
                    state.isMissing = state.day == nil && !dataState.isLoading
                }
                state.isLoading = dataState.isLoading
                // A refresh failure with a day already resolved is not worth a banner here: the
                // hourly readings for a single past-or-present day do not change materially.
                state.error = state.day == nil ? dataState.error : nil
            }
        } catch {
            state.isLoading = false
            state.error = AppError.from(error)
        }
    }
}

/// The 3-hourly breakdown for one forecast day, pushed from the Forecast tab.
struct DayDetailScreen: View {
    let locationID: Int64
    let date: Date

    @Environment(AppContainer.self) private var container
    @State private var viewModel: DayDetailViewModel?

    var body: some View {
        Group {
            if let viewModel {
                DayDetailContent(state: viewModel.state)
            } else {
                LoadingView()
            }
        }
        // The forecast location's zone, matching the row that pushed this screen, otherwise the
        // title can name a different weekday from the row the user tapped.
        .navigationTitle(
            date.formatted(
                Date.FormatStyle(timeZone: viewModel?.state.timeZone ?? .gmt)
                    .weekday(.wide)
                    .day()
                    .month(.abbreviated)
            )
        )
        .navigationBarTitleDisplayMode(.inline)
        .task {
            if viewModel == nil {
                viewModel = DayDetailViewModel(
                    locationID: locationID,
                    date: date,
                    weatherRepository: container.weatherRepository,
                    locationRepository: container.locationRepository,
                    settingsStore: container.settingsStore
                )
            }
            viewModel?.start()
        }
        .onDisappear { viewModel?.stop() }
    }
}

/// The stateless half.
struct DayDetailContent: View {
    let state: DayDetailUiState

    var body: some View {
        if state.isMissing {
            EmptyStateView(
                title: "Day not available",
                message: "This day has rolled out of the five-day forecast window.",
                systemImage: "calendar.badge.exclamationmark"
            )
        } else if let day = state.day {
            list(day)
        } else if let error = state.error {
            // No retry: the Forecast tab owns refreshing, and this screen has no location to
            // refresh against until one resolves.
            ErrorView(error: error)
        } else {
            LoadingView()
        }
    }

    private func list(_ day: ForecastDay) -> some View {
        List {
            Section {
                DaySummary(day: day, unit: state.preferences.temperatureUnit)
                    .listRowBackground(Color.clear)
            } header: {
                Text(state.locationName)
            }

            Section("Hour by hour") {
                ForEach(day.hourly) { hour in
                    HourlyRow(
                        hour: hour,
                        timeZone: state.timeZone,
                        temperatureUnit: state.preferences.temperatureUnit,
                        windUnit: state.preferences.windSpeedUnit
                    )
                }
            }
        }
    }
}

/// The day's headline: condition, description and high/low.
private struct DaySummary: View {
    let day: ForecastDay
    let unit: TemperatureUnit

    private var high: Int {
        Int(unit.convertFromCelsius(day.maxTemperatureCelsius).rounded())
    }

    private var low: Int {
        Int(unit.convertFromCelsius(day.minTemperatureCelsius).rounded())
    }

    var body: some View {
        VStack(spacing: Spacing.sm) {
            ConditionBadge(condition: day.condition, isDaytime: true)
            Text(day.description)
                .font(.title3.weight(.semibold))
            Text("High \(high)\(unit.symbol) · Low \(low)\(unit.symbol)")
                .font(.subheadline)
                .foregroundStyle(.secondary)
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, Spacing.sm)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(
            "\(day.description), high \(high)\(unit.symbol), low \(low)\(unit.symbol)"
        )
    }
}

/// One 3-hourly reading.
private struct HourlyRow: View {
    let hour: HourlyForecast
    /// The forecast location's zone, not the device's, or a Maldivian forecast read from London
    /// would list its afternoon readings as morning ones.
    let timeZone: TimeZone
    let temperatureUnit: TemperatureUnit
    let windUnit: WindSpeedUnit

    private var time: String {
        hour.time.formatted(
            Date.FormatStyle(date: .omitted, time: .shortened, timeZone: timeZone)
        )
    }

    private var temperature: Int {
        Int(temperatureUnit.convertFromCelsius(hour.temperatureCelsius).rounded())
    }

    private var wind: String {
        let value = windUnit.convertFromMetresPerSecond(hour.windSpeedMetresPerSecond)
        return "\((value * 10).rounded() / 10) \(windUnit.symbol)"
    }

    private var rainPercent: Int {
        Int((hour.precipitationProbability * 100).rounded())
    }

    var body: some View {
        HStack(spacing: Spacing.md) {
            Text(time)
                .font(.subheadline)
                .monospacedDigit()
                // A fixed width keeps the temperature column aligned down the list; without it
                // "9 AM" and "12 PM" push everything sideways.
                .frame(width: timeColumnWidth, alignment: .leading)

            Image(systemName: hour.condition.symbolName(isDaytime: isDaytime))
                .symbolRenderingMode(.multicolor)
                .accessibilityHidden(true)

            Text("\(temperature)\(temperatureUnit.symbol)")
                .font(.headline)
                .monospacedDigit()

            Spacer(minLength: Spacing.sm)

            VStack(alignment: .trailing, spacing: Spacing.xxs) {
                Text(wind)
                    .font(.caption)
                if rainPercent > 0 {
                    Text("\(rainPercent)%")
                        .font(.caption)
                        .foregroundStyle(Color.skyAccent)
                }
            }
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(announcement)
    }

    /// The forecast carries no sunrise/sunset, so daylight is approximated by clock hour. Getting
    /// this wrong would show a sun at 3 am, the exact parity bug `WeatherConditionIconTest`
    /// guards against, so it is worth being explicit rather than passing `isDaytime: true`.
    private var isDaytime: Bool {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let hourOfDay = calendar.component(.hour, from: hour.time)
        return hourOfDay >= dawnHour && hourOfDay < duskHour
    }

    private var announcement: String {
        var text = "\(time), \(temperature)\(temperatureUnit.symbol), wind \(wind)"
        if rainPercent > 0 {
            text += ", \(rainPercent) percent chance of rain"
        }
        return text
    }

    private let timeColumnWidth: CGFloat = 64
    private let dawnHour = 6
    private let duskHour = 20
}
